packages/protocol/contracts/L1/v1/V1Proving.sol

```diff
@@ -34,7 +32,7 @@ library V1Proving {
         LibData.BlockMetadata meta;
         BlockHeader header;
         address prover;
-        bytes[] proofs;
+        bytes[] proofs; // The first K_ZKPROOFS_PER_BLOCK are ZKPs
```

Comment:
With recursion it's also possible to just verify the different proofs in another proof so onchain we still only see a single proof (with perhaps some extra data so the smart contract knows which proofs were valid), this way there is less work to be done on-chain. In the proof the recursion could also be done using different recursion implementations to also guard against bugs inside a single implementation of recursion. I guess the difference in these methods is small when we will allow provers to just post the data of the proof on-chain and let other provers batch verify them, though proofs may get quite large under HyperPlonk/plonky2 so fewer proofs on-chain could be quite a bit more efficient.

Anyway, a fully on-chain method for now makes sense to me because it remains to be soon how practical using recursion for this will even be.

Reply:
If the aggregation circuits can be audited more carefully so we know it's bug-free, this approach will be a better one as it reduces gas-cost.